Fix display tests and enable when possible

Martin-Idel-SI opened the pull request:
This is a followup for #175 . It aims to fix all display tests and run them when possible.

- In order to not introduce visibility control in `rviz_default_plugins`, we need to build the library both static and shared. Apparently, includes are handled differently in that way, as the static building generates a lot of OGRE and pluginlib warnings on both Windows and Linux.
- To run tests, we try to decide whether OpenGL exists via CMake.

greimela-si committed:
Enable display tests on Linux only if display variable is available

Also enables tests that require a display on Mac and Windows.

Martin-Idel-SI committed:
Fix Windows warnings in tests

Martin-Idel-SI committed:
Introduce pragmas to fix Linux warnings

Martin-Idel-SI committed:
Fix display tests needing ogre_media

Martin-Idel-SI commented:
First try with CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3912)](http://ci.ros2.org/job/ci_linux/3912/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1031)](http://ci.ros2.org/job/ci_linux-aarch64/1031/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3236)](http://ci.ros2.org/job/ci_osx/3236/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4007)](http://ci.ros2.org/job/ci_windows/4007/)

Martin-Idel-SI commented:
- So it seems that disabling the tests if no display is available on Linux works. 
- I would assume that the SEH errors on Windows indicate there is no display available or there is an OpenGL problem. The others could just be a different symptom. Can you give more information about the Windows Jenkins @wjwwood ?
- On Mac, there seems to be a different problem, I'll try to reproduce it locally tomorrow and see how far this goes.

wjwwood commented on rviz_default_plugins/src/rviz_default_plugins/displays/camera/camera_display.cpp:
```diff
@@ -545,5 +553,13 @@ void CameraDisplay::reset()
+# pragma GCC diagnostic push
+# pragma GCC diagnostic ignored "-Wunused-parameter"
+# pragma GCC diagnostic ignored "-Wpedantic"
+#endif
```
Why is this needed?

wjwwood replied:
We should only disable warnings like this when they're in external dependencies. This looks like an issue within pluginlib?

Martin-Idel-SI replied:
I totally agree, I just wanted to see whether I could get it to compile if I did this.

I get the following warning: 
```
/include/pluginlib/class_list_macros.hpp:49:59: error: extra ‘;’ [-Werror=pedantic]
   CLASS_LOADER_REGISTER_CLASS(class_type, base_class_type);
```
If I read this correctly, it complains that the ";" is superfluous in this line:

https://github.com/ros/pluginlib/blob/b456beba88564b746694dde68696d8bc4edc78c9/pluginlib/include/pluginlib/class_list_macros.hpp#L49

It seems to me that compiling the pluginlib with -Werror=pedantic should catch that problem.

wjwwood replied:
Looks like this comment got moved to a new line of code or something... Right now it appears to only apply to Ogre includes, which is fine.

Just to reiterate, I'd say it's fine to suppress `-Wpedantic` for `pluginlib`, but not unused variables. That should instead be fixed upstream. Ideally, we'd use pedantic upstream in pluginlib too.

wjwwood commented on rviz_default_plugins/src/rviz_default_plugins/displays/image/ros_image_texture.cpp:
```diff
@@ -40,7 +40,15 @@
+# pragma GCC diagnostic ignored "-Wunused-parameter"
+# pragma GCC diagnostic ignored "-Wpedantic"
+#endif
 #include <OgreTextureManager.h> // NOLINT: cpplint cannot handle include order
```
nitpick: please add a second space between the code and comment

wjwwood commented on rviz_default_plugins/src/rviz_default_plugins/displays/pointcloud/point_cloud_display.cpp:
```diff
@@ -31,9 +31,15 @@
 
 #include <memory>
 #include <utility>
-
```
nitpick: leave the space here

wjwwood commented on rviz_default_plugins/src/rviz_default_plugins/displays/pointcloud/point_cloud_display.cpp:
```diff
@@ -31,9 +31,15 @@
-
+#ifndef _WIN32
+# pragma GCC diagnostic pop
+#endif
```
Put a space after this.

wjwwood reviewed (commented):
I have a few nitpicks, I know it's still a WIP, but I thought I'd give some early feedback.

Martin-Idel-SI committed:
Fix pragmas for Apple and add a few spaces

Martin-Idel-SI commented:
Thanks for the early review! Here is a new CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3925)](http://ci.ros2.org/job/ci_linux/3925/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1041)](http://ci.ros2.org/job/ci_linux-aarch64/1041/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3250)](http://ci.ros2.org/job/ci_osx/3250/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4019)](http://ci.ros2.org/job/ci_windows/4019/)

Looking into the output, display tests can apparently run on Mac! 
What's left to do?
- I guess the Apple failures are due to missing environment variables? So far, I couldn't reproduce locally, but I've never built isolated, so I might get to it tomorrow
- There are a lot of different failures on Windows, but since all tests fail, I'd say the display is not present. However, it might be a problem of isolated builds vs non-isolated builds. I'll investigate that one tomorrow.

Martin-Idel-SI committed:
Add path to fix rviz_rendering_tests

Martin-Idel-SI commented:
The failures on Mac (and some on Windows) are due to environment problems. The last commit fixes the ones for Mac:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3929)](http://ci.ros2.org/job/ci_linux/3929/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1047)](http://ci.ros2.org/job/ci_linux-aarch64/1047/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3256)](http://ci.ros2.org/job/ci_osx/3256/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4025)](http://ci.ros2.org/job/ci_windows/4025/)

For Windows, I haven't tried, yet - but at least we now know that we can run all tests on Mac!

wjwwood commented:
Looks like they are now running, but failing on Windows.

I'd expect our macOS and Windows machines to work since they are all dedicated machines and not headless servers like our linux machines.

Martin-Idel-SI commented:
Yes, I expected the tests run successfully on Windows, too. I'm trying to reproduce the failures locally but so far I've been unsuccessful. I'll work on that next week after fixing the issues in the release candidate.

Martin-Idel-SI committed:
Fix Polygon display

Martin-Idel-SI committed:
Add check to fix display tests

- initialize must have been called before the window exists
- this can only happen when rendering

Martin-Idel-SI commented:
I did a rebase and had a closer look again. 
I tried to reproduce the errors on our Windows machine in several ways, but couldn't
- Just build rviz locally in its own repository (clean console, ros2 sourced)
- Build rviz isolated in its own repository (clean console, ros2 sourced)
- Build ros2 with this rviz branch locally using --build-tests --isolated (clean console, no source) and then ament.py test as on your CI farm. 
All tests always ran. I just can't reproduce.

There are many different kinds of errors. I've seen most of them before in some version - they turned out to be mostly environment related.

Some observations I made:
- For the tests in rviz_default_plugins, the failure with some big number is very probably due to dlls not being found
- The tests for PointCloud2Display run, so it seems that the Qt dlls are found
- We once had an issue when curl wasn't found, this might be the case
- For rviz_rendering tests, it seems that ogre can be found, but doesn't work correctly. It can't be a problem of the display being locked (this is a problem with GUI tests), we tested that
- Including AMENT_PREFIX_PATH and PATH doesn't really change anything in the errors produced (see the errors in point_cloud_renderable_test, where I added the paths and billboard_line_test, where I didn't), so it doesn't seem to be a problem with those parts of the environment

What I could try is to get CMake to print out all environment variables and hope that I can then tweak the environment locally to reproduce the test failures, but that would probably mean quite a few jobs on ros2 CI. Do you have any better suggestion?

Martin-Idel-SI committed:
Disable display tests per default on Win32

- Offer a CMake variable to enable them via commandline

Martin-Idel-SI commented:
I can't reproduce the errors locally, whatever I do. That means it's really difficult to debug, so for now, I disabled the tests on Windows. They can be reenabled sending a CMake flag, so one could easily debug the problem when having access to the physical machines.

It remains to fix the pluginlib so that we don't need the the pragmas for that.

CI with display tests disabled on Windows:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3953)](http://ci.ros2.org/job/ci_linux/3953/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1072)](http://ci.ros2.org/job/ci_linux-aarch64/1072/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3287)](http://ci.ros2.org/job/ci_osx/3287/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4056)](http://ci.ros2.org/job/ci_windows/4056/)

The failed tests don't have anything to do with this.

greimela-si commented:
Rebased after changes on ros2 branch.

New CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4022)](http://ci.ros2.org/job/ci_linux/4022/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1116)](http://ci.ros2.org/job/ci_linux-aarch64/1116/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3349)](http://ci.ros2.org/job/ci_osx/3349/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4113)](http://ci.ros2.org/job/ci_windows/4113/)

Martin-Idel-SI committed:
Fix rviz_rendering includes

Martin-Idel-SI committed:
Fix pragmas for selection_manager

Martin-Idel-SI committed:
Fix pragmas in robot_model_display

Martin-Idel-SI committed:
Fix include problems for urdf and tinyxml

Martin-Idel-SI commented on rviz_common/CMakeLists.txt:
```diff
@@ -280,6 +283,7 @@ ament_export_dependencies(
   tf2_geometry_msgs
   tf2_ros
 )
+ament_export_include_directories(include)
```
What I don't understand is why this suddenly becomes necessary. If I only build dynamically, the call to `ament_export_interfaces(rviz_common)` is enough (although the `rviz_common_INCLUDE_DIRS` are not properly populated), but once I try to build the object files, I can't find include files from `rviz_rendering`. 
I can fix this here no problem, but I have the same issue with resource_retriever, where I can't add the call to `ament_export_include_directories`.
Sorry to bother you @wjwwood , but can you explain this phenomenon?

wjwwood replied:
`ament_export_interfaces(rviz_common)` exports the cmake target `rviz_common`, so when using it in another package (as `rviz_common::rviz_common`), it brings with it the include dirs, link flags, etc...

In this mode of operation (target exporting) all settings are target specific, so things like `rviz_common_INCLUDE_DIRS` are not used.

If you want that directory to be included in the `rviz_common` target you need some `target_include_directories` magic, see:

https://github.com/bsinno/rviz/blob/6f27b4dfa42b1adc6c8692fd866600fb5282b32d/rviz_common/CMakeLists.txt#L236-L242

Which is already happening, so you should only need to link what ever you're compiling that needs those headers against `rviz_common::rviz_common` to get the right stuff.

Martin-Idel-SI replied:
Thank you for the explanation. 
That's actually a problem, because if you compile OBJECT libraries like I do in `rviz_default_plugins` (to later compile to shared libraries for the real application and to a static library to link against tests), you can't specify link targets. 

More specifically: In CMakeLists of `rviz_default_plugins`, I'm doing
```
add_library(rviz_default_plugins_object OBJECT
  ${rviz_default_plugins_headers_to_moc}
  ${rviz_default_plugins_source_files}
)
```
If I don't have the `ament_export_include_directories` in `rviz_common`, I can't find include directories of `resource_retriever` specified via `target_include_directories(rviz_default_plugins_object PUBLIC ${rviz_common_INCLUDE_DIRS})`
what I'd have to do is say
`target_link_libraries(rviz_default_plugins_object PUBLIC rviz_common::rviz_common)`
or 
`ament_target_dependencies(rviz_default_plugins_object rviz_common)`.
However, neither is possible, because you mustn't link against OBJECT libraries.
That explains why I need a `ament_export_include_directories`. For `rviz_common` that's no problem, for `resource_retriever` I can't do this, but I asked whether this could be added (https://github.com/ros/resource_retriever/issues/21).

Martin-Idel-SI commented:
I'm trying to rebase on master, but I'm having trouble with the new resource_retriever dependency in rviz_default_plugins.

wjwwood commented:
Also, with respect to the Windows issues, one problem might be that we run the jenkins jobs as the `SYSTEM` user. You can get a cmd prompt as this user by using the `psexec` tool, see: https://blogs.technet.microsoft.com/askds/2008/10/22/getting-a-cmd-prompt-as-system-in-windows-vista-and-windows-server-2008/

Martin-Idel-SI committed:
Introduce pragmas for pose display

Martin-Idel-SI commented:
Using the system user might of course change a few things. Maybe I can reproduce the failures with that.

Martin-Idel-SI committed:
Remove pragmas from pluginlib imports

Martin-Idel-SI committed:
Fix AMENT_PREFIX_PATH for robot_model_test

Martin-Idel-SI commented:
Let's see where we are at:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4063)](http://ci.ros2.org/job/ci_linux/4063/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1163)](http://ci.ros2.org/job/ci_linux-aarch64/1163/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3373)](http://ci.ros2.org/job/ci_osx/3373/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4138)](http://ci.ros2.org/job/ci_windows/4138/)

Still missing: I haven't enabled the tests on Windows, yet. Building as system user fails atm due to other problems, so I couldn't reproduce the errors yet.

If we want to fix that later, we should be "good to go" (if CI agrees).

Martin-Idel-SI committed:
Disable rviz_rendering_tests on Windows

Martin-Idel-SI committed:
Fix clang warnings by introducing pragmas

- Warnings only concern Ogre

Martin-Idel-SI commented:
I had to disable a few more tests and introduce some Mac pragmas:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4069)](http://ci.ros2.org/job/ci_linux/4069/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1172)](http://ci.ros2.org/job/ci_linux-aarch64/1172/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3376)](http://ci.ros2.org/job/ci_osx/3376/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4141)](http://ci.ros2.org/job/ci_windows/4141/)

greimela-si committed:
Add missing pragmas for marker and pose array display

Martin-Idel-SI commented:
CI after rebase:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4154)](http://ci.ros2.org/job/ci_linux/4154/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1235)](http://ci.ros2.org/job/ci_linux-aarch64/1235/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3442)](http://ci.ros2.org/job/ci_osx/3442/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4259)](http://ci.ros2.org/job/ci_windows/4259/)

Martin-Idel-SI committed:
Fix MarkerDisplay pragma for Apple

Martin-Idel-SI commented:
@wjwwood I tried to investigate some more, but the system user is no problem.

I guess it's may be a combination of having a screen lock, the system user and maybe some other things. In that case, I would be surprised if there was a way to make the tests pass on Windows at all. 

In any case, this is where we are:
- Linux: Tests run if there is a display. For instance, enabling an xvfb plugin for Jenkins or having a PC automatically enables them. 
- Windows: Tests only run if you explicitly tell them to. 
- Apple: Tests run.

The only downside is some cmake code to make it work that way.
Maybe we should merge the branch as is? That would make our lives easier and your ci also runs more tests and verifies that things work at least on one target system. When a new idea comes up, we can also discuss it.

CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4156)](http://ci.ros2.org/job/ci_linux/4156/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1237)](http://ci.ros2.org/job/ci_linux-aarch64/1237/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3450)](http://ci.ros2.org/job/ci_osx/3450/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4261)](http://ci.ros2.org/job/ci_windows/4261/)

wjwwood commented:
> Maybe we should merge the branch as is? That would make our lives easier and your ci also runs more tests and verifies that things work at least on one target system. When a new idea comes up, we can also discuss it.

In general, I'm in favor of this, I'll resolve the conflicts and do a code review.

wjwwood commented:
Actually I don't think the conflict resolution is trivial since merging the other pr's. Sorry, I didn't mean to cause a problem with the merge order. I think, however, that if this were merged first several other pr's would need to be de-conflicted. Not sure there would have been a better order.

Can you have a look at the conflicts in the `rviz_default_plugins/CMakeLists.txt`?

wjwwood commented on rviz_common/CMakeLists.txt:
```diff
@@ -337,6 +331,26 @@ if(BUILD_TESTING AND NOT WIN32)
+    ament_add_gtest(rviz_common_display_test
+      test/display_test.cpp
+      test/mock_display.cpp
+      test/mock_display_group.cpp)
```
For all of these, maybe it would be better to use the `SKIP_TEST` argument of the `ament_add_*test()` macros:

https://github.com/ament/ament_cmake/blob/f74ccfa7399156d20c8d0eff8100445b7eefebd9/ament_cmake_gtest/cmake/ament_add_gtest.cmake#L37-L38

I usually do something like this:

```cmake
set(SHOULD_SKIP_DISPLAY_TESTS "")
if(<condition, like is there a display>)
  # if we should not run tests
  set(SHOULD_SKIP_DISPLAY_TESTS SKIP_TEST)
endif()

...

ament_add_gtest(<target> <src1> <src2> ... ${SHOULD_SKIP_DISPLAY_TESTS})
```

The difference is that it will show up as a skipped test in our test results rather than not appearing at all, e.g.:

https://ci.ros2.org/view/nightly/job/nightly_win_rel/772/testReport/composition.build.composition/test_linktime_composition__rmw_connext_cpp_Release/

greimela-si replied:
Yes, that would be nice. 

However, we are using `ament_add_gmock()` for most of the tests and this function does not support the `SKIP_TEST` argument.

greimela-si replied:
I created an issue for it: https://github.com/ament/ament_cmake/issues/129

wjwwood commented on rviz_default_plugins/package.xml:
```diff
@@ -26,10 +26,12 @@
   <depend>rviz_rendering</depend>
   <depend>urdf</depend>
   <depend>visualization_msgs</depend>
+  <depend>tinyxml_vendor</depend>
```
nitpick: alphabetical order

wjwwood commented on rviz_default_plugins/src/rviz_default_plugins/tools/move/move_tool.cpp:
```diff
@@ -81,5 +81,13 @@ int MoveTool::processKeyEvent(QKeyEvent * event, rviz_common::RenderPanel * pane
+# pragma GCC diagnostic push
+# pragma GCC diagnostic ignored "-Wunused-parameter"
+# pragma GCC diagnostic ignored "-Wpedantic"
+#endif
```
Ah, here we go, can the `-Wunused-parameter` be removed? I feel like that should be checked in `pluginlib` already.

If not, I'll make a pr to `pluginlib` to fix both of these.

greimela-si replied:
You are right, the `-Wunused-parameter` is obsolete.

wjwwood commented on rviz_default_plugins/test/rviz_default_plugins/displays/image/ros_image_texture_test.cpp:
```diff
@@ -40,6 +44,9 @@
 #include "../../../../src/rviz_default_plugins/displays/image/ros_image_texture.hpp"
+#ifdef _WIN32
+# pragma warning(pop)
+#endif
```
This catch is too broad. It should be at the location where we include an external header. This covers system libraries, Ogre, gtest and gmock (should only use one of those too right?), generated code, and our own headers.

greimela-si replied:
Looks like this catch is not needed at all.

wjwwood reviewed (changes requested):
I asked for a few small changes, but overall, I agree we should merge and then improve it (after the small changes and merge conflicts are resolved).

greimela-si committed:
Add missing pragmas and fix cmake

greimela-si committed:
Fix tests

greimela-si committed:
Add laser_geometry include dir

greimela-si committed:
Add nav_msgs include dirs

greimela-si committed:
Fix PATH in robot_test

greimela-si committed:
Remove obsolete pragmas

greimela-si commented:
I fixed all the merge conflicts and included your review comments.

New CI run:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4179)](http://ci.ros2.org/job/ci_linux/4179/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1248)](http://ci.ros2.org/job/ci_linux-aarch64/1248/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3460)](http://ci.ros2.org/job/ci_osx/3460/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4276)](http://ci.ros2.org/job/ci_windows/4276/)

wjwwood reviewed (approved):
lgtm

dirk-thomas commented on rviz_default_plugins/CMakeLists.txt:
```diff
@@ -108,14 +115,46 @@ set(rviz_default_plugins_source_files
+  ${Qt5Widgets_INCLUDE_DIRS}
+  ${resource_retriever_INCLUDE_DIRS}
+  ${TinyXML_INCLUDE_DIRS}
+  ${urdf_INCLUDE_DIRS}
```
This patch broke the build again when `urdf` is installed on the system. See #243 for a temp. workaround.
